### PR TITLE
cache boxed integers in a radix tree, make HTTP status and port primitives

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

**Comment by richardstartin** [comment]
There's a small decrease in old gen (JDK8/PGC) in spring petclinic, but hard to justify the change really. I haven't attempted to measure CPU usage.

image
image

That said, it's a very small and contained change.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/cache/RadixTreeBoxCache.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/** Sparse cache of commonly boxed values */
+public final class RadixTreeBoxCache {
+
+  public static final RadixTreeBoxCache HTTP_STATUSES =
+      new RadixTreeBoxCache(8, 64, 200, 201, 301, 307, 400, 401, 403, 404, 500, 502, 503);
+
+  public static final RadixTreeBoxCache PORTS =
+      new RadixTreeBoxCache(256, 256, 80, 443, 8080, 9090);
```
Comment (tylerbenson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/cache/RadixTreeBoxCache.java:12]: Is `9090` common enough to warrant caching here?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/cache/RadixTreeBoxCache.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package datadog.trace.bootstrap.instrumentation.cache;
+
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/** Sparse cache of commonly boxed values */
+public final class RadixTreeBoxCache {
+
+  public static final RadixTreeBoxCache HTTP_STATUSES =
+      new RadixTreeBoxCache(8, 64, 200, 201, 301, 307, 400, 401, 403, 404, 500, 502, 503);
```
Comment (tylerbenson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/cache/RadixTreeBoxCache.java:9]: A comment with how you determined your l1 and l2 sizes would be helpful.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/cache/RadixTreeBoxCache.java (hunk 1 above)
Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/cache/RadixTreeBoxCache.java:12]: Probably not.

**Review by bantonsson — APPROVED** [review]
Nice

**Pull request merged**